### PR TITLE
Global Liquid Staking Cap

### PR DESCRIPTION
**Pull request by sampocs**
## Context
As an additional safety measure for liquid staking, this PR introduces a global liquid staking cap. For instance, a cap of 0.3 means only 30% of staked ATOM can be liquid staked. The liquid staking volume is determined by tracking tokenized shares and/or delegations from ICA accounts. The total stake is calculated from the total balance of the Bonded pool.

[ADR](https://github.com/iqlusioninc/liquidity-staking-module/blob/riley/adr-2/adr/adr-002.md) (courtesy of @riley-stride) 

**Action Item**: There is still 1 small required change to the evidence module to handle tombstoned validators (see ADR/Spec)

## Brief Changelog
* Added `global_liquid_staking_cap` parameter
* Added keeper functions to track the total liquid staked across all delegations
* Added helper functions to track the global liquid stake and the validator’s liquid shares
* Added error cases and accounting changes to Delegate/Undelegate/Redelegate/TokenizeShares/RedeemTokens to accommodate the global and validator caps
* Fixed broken `ValidatorBond` unit test and added additional test cases

### CODE REVIEW

**Comment by faddat** [comment]
moment on this one -- I am going to do a bunch of branch tidying and get this in.


Thank you so much.

**Comment by faddat** [comment]
@sampocs -- I am not 100% sure about this one.

I think that I may prefer a tax to a cap.

wdyt?

File: x/staking/keeper/msg_server.go [hunk 1]
```diff
@@ -383,17 +411,18 @@ func (k msgServer) Undelegate(goCtx context.Context, msg *types.MsgUndelegate) (
...
 		}
+	}
 
-		// reduce total validator-bond delegation on unbond
-		validator.TotalValidatorBondShares = validator.TotalValidatorBondShares.Sub(shares)
-		k.SetValidator(ctx, validator)
+	// if this undelegation is from a liquid staking provider, the global and validator
+	// liquid counts should be decremented
+	if k.AccountIsLiquidStakingProvider(ctx, delegatorAddress) {
+		k.DecreaseTotalLiquidStakedTokens(ctx, tokens)
```
Comment (riley-stride) [inline x/staking/keeper/msg_server.go:424]: Do you think we should emit some events when LS tokens are incremented/decremented, and when we perform other important LS related state changes?

File: x/staking/keeper/msg_server.go [hunk 2]
```diff
@@ -688,10 +721,6 @@ func (k msgServer) TokenizeShares(goCtx context.Context, msg *types.MsgTokenizeS
 		return nil, err
 	}
 
-	validator, _ = k.GetLiquidValidator(ctx, valAddr)
```
Comment (riley-stride) [inline x/staking/keeper/msg_server.go:691]: When swapping out the previous LSM logic (here) with new LSM logic (above), why did you move it earlier in the `RedeemTokens` execution path? Same question for the other x/staking keeper function changes

File: x/staking/keeper/liquid_stake_test.go [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+// Tests AccountIsLiquidStakingProvider
+func TestAccountIsLiquidStakingProvider(t *testing.T) {
+	_, app, ctx := createTestInput(t)
+
+	// Create base account
+	baseAccountAddress := sdk.AccAddress("base-account")
+	app.AccountKeeper.SetAccount(ctx, authtypes.NewBaseAccountWithAddress(baseAccountAddress))
+
+	// Create an ICA module account
+	icaModuleAccountName := "ica-account"
```
Comment (riley-stride) [inline x/staking/keeper/liquid_stake_test.go:58]: can we also test a non-ICA module account?

File: x/staking/keeper/liquid_stake.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	return tokens
+}
+
+// Check if an account is a owned by a liquid staking provider
+// This is determined by checking if the account is a 32-length module account
+func (k Keeper) AccountIsLiquidStakingProvider(ctx sdk.Context, address sdk.AccAddress) bool {
+	account := k.authKeeper.GetAccount(ctx, address)
+	_, isModuleAccount := account.(*authtypes.ModuleAccount)
```
Comment (riley-stride) [inline x/staking/keeper/liquid_stake.go:46]: Can we make sure we have thorough testing on this for the following cases
- non-ICA module accounts (e.g. staking module accounts)
- interchain accounts
- regular private key accounts
- we can't test CW account afaiu on Gaia (note to self to test those before generalizing LSM to other zones)

Can you think of other types?

File: x/staking/keeper/liquid_stake_test.go [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+		TotalLiquidShares: initialLiquidShares,
+	}
+	app.StakingKeeper.SetValidator(ctx, initialValidator)
+
+	// Set validator bond factor so that the check is disabled
+	params := app.StakingKeeper.GetParams(ctx)
+	params.ValidatorBondFactor = sdk.NewDec(-1)
+	app.StakingKeeper.SetParams(ctx, params)
+
+	// Attempt to decrease the validator liquid shares, it should not be changed since the cap is disabled
```
Comment (riley-stride) [inline x/staking/keeper/liquid_stake_test.go:499]: What's the test doing here? 

Are you doing the following?
- Start with some liquid shares
- Disable cap
- Try to change liquid shares -> should fail because cap is turned off

File: x/staking/keeper/liquid_stake_test.go [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+
+	err := app.BankKeeper.MintCoins(ctx, minttypes.ModuleName, sdk.NewCoins(bondedPoolCoin))
+	require.NoError(t, err, "no error expected when minting")
+
+	err = app.BankKeeper.SendCoinsFromModuleToModule(ctx, minttypes.ModuleName, types.BondedPoolName, sdk.NewCoins(bondedPoolCoin))
+	require.NoError(t, err, "no error expected when sending tokens to bonded pool")
+}
+
+// Tests CheckExceedsGlobalLiquidStakingCap
+func TestCheckExceedsGlobalLiquidStakingCap(t *testing.T) {
```
Comment (riley-stride) [inline x/staking/keeper/liquid_stake_test.go:93]: Could we make sure we have at least these three cases covered here:
1) No LSM tokenizations, only ICA account delegations. Exceeds cap.
2) ICA account delegations, only LSM tokenizations from non-module account. Exceeds cap.
3) Some LSM tokenizations, some ICA account delegations. Exceeds cap.
Should give us confidence both cases are individually covered and the sum too

File: x/staking/keeper/liquid_stake_test.go [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+
+	// Create a validator with designated self-bond shares
+	privKey := secp256k1.GenPrivKey()
+	pubKey := privKey.PubKey()
+	valAddress := sdk.ValAddress(pubKey.Address())
+
+	initialValidator := types.Validator{
+		OperatorAddress:          valAddress.String(),
+		TotalLiquidShares:        initialLiquidShares,
+		TotalValidatorBondShares: validatorBondShares,
```
Comment (riley-stride) [inline x/staking/keeper/liquid_stake_test.go:435]: I see in these tests we set the ValidatorBondShares directly in state. The other PR tests the process of delegating -> marking delegation as ValidatorBond ?

**Review by riley-stride — CHANGES_REQUESTED** [review]
Looking really good @sampocs! Added minor change requests and a few question, if you prefer to walk through it on a call just lmk

File: x/staking/spec/08_params.md [hunk 8]
```diff
@@ -6,12 +6,13 @@ order: 8
...
+| Key                    | Type             | Example                |
+| ---------------------- | ---------------- | ---------------------- |
+| UnbondingTime          | string (time ns) | "259200000000000"      |
+| MaxValidators          | uint16           | 100                    |
+| KeyMaxEntries          | uint16           | 7                      |
+| HistoricalEntries      | uint16           | 3                      |
+| BondDenom              | string           | "stake"                |
+| MinCommissionRate      | string           | "0.000000000000000000" |
+| ValidatorBondFactor    | string           | "250.0000000000000000" |
+| GlobalLiquidStakingCap | string           | "0.300000000000000000" |
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/staking/spec/08_params.md:18]: done

File: x/staking/keeper/msg_server.go [hunk 9]
```diff
@@ -257,10 +274,25 @@ func (k msgServer) Delegate(goCtx context.Context, msg *types.MsgDelegate) (*typ
...
 	valSrcAddr, err := sdk.ValAddressFromBech32(msg.ValidatorSrcAddress)
 	if err != nil {
 		return nil, err
 	}
+	valDstAddr, err := sdk.ValAddressFromBech32(msg.ValidatorDstAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	srcValidator, found := k.GetLiquidValidator(ctx, valSrcAddr)
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/staking/keeper/msg_server.go:287]: I was wondering the same thing - would love to hear @zmanian or @faddat's take on it

Imo, when this gets merged into the SDK, we should just have one function here (same with `GetDelegation`), but I assumed it was setup this way intentionally - maybe because other modules call `GetValidator` so we need to keep the legacy function?

Here's the [commit](https://github.com/iqlusioninc/liquidity-staking-module/commit/dc4e1ac50f83cafb3fcd03b74fb0ca10737385dc) - I assume there was good reason for it

File: x/staking/keeper/liquid_stake.go [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+// GetTotalLiquidStakedTokens returns the total outstanding tokens owned by a liquid staking provider
+// Returns zero if the total liquid stake amount has not been initialized
+func (k Keeper) GetTotalLiquidStakedTokens(ctx sdk.Context) sdk.Int {
+	store := ctx.KVStore(k.storeKey)
+	tokensBz := store.Get(types.TotalLiquidStakedTokensKey)
+
+	if tokensBz == nil {
+		// QUESTION: Should we panic here instead?
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/staking/keeper/liquid_stake.go:28]: I'd prefer panic'ing over throwing an error. I actually think it's pretty [common practice ](https://github.com/cosmos/cosmos-sdk/blob/main/x/staking/keeper/delegation.go) to just panic in these keeper functions since they usually indicate that something is very wrong (different than a typical error that might arise from invalid input). 

I think the question is whether to panic or just return 0. I think I probably lean panic so we don't accidentally throw off the accounting

File: x/staking/keeper/liquid_stake.go (hunk 4 above)
Comment (sampocs) [reply x/staking/keeper/liquid_stake.go:46]: I'm not sure how to differentiate a non-ICA module account from a module account in this scenario - there's no additional identifier besides an arbitrary address and name

File: x/staking/keeper/liquid_stake.go [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+	return liquidStakePercent.GT(liquidStakingCap)
+}
+
+// ExceedsValidatorBondCap checks if a liquid delegation to a validator would cause
+// the liquid shares to exceed the validator bond factor
+// A liquid delegation is defined as either tokenized shares, or a delegation from an ICA Account
+// Returns true if the cap is exceeded
+func (k Keeper) CheckExceedsValidatorBondCap(ctx sdk.Context, validator types.Validator, shares sdk.Dec) bool {
+	validatorBondFactor := k.ValidatorBondFactor(ctx)
+	maxValTotalShare := validator.TotalValidatorBondShares.Mul(validatorBondFactor)
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/staking/keeper/liquid_stake.go:77]: good call! done

File: x/staking/keeper/liquid_stake.go [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+	k.SetTotalLiquidStakedTokens(ctx, k.GetTotalLiquidStakedTokens(ctx).Add(amount))
+
+	return nil
+}
+
+// DecreaseTotalLiquidStakedTokens decrements the total liquid staked tokens
+// if the global cap is enabled
+func (k Keeper) DecreaseTotalLiquidStakedTokens(ctx sdk.Context, amount sdk.Int) {
+	if k.GlobalLiquidStakingCapEnabled(ctx) {
+		k.SetTotalLiquidStakedTokens(ctx, k.GetTotalLiquidStakedTokens(ctx).Sub(amount))
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/staking/keeper/liquid_stake.go:104]: shouldn't be possible, but we can floor it at zero if you think it's a concern

File: x/staking/keeper/msg_server.go (hunk 1 above)
Comment (sampocs) [reply x/staking/keeper/msg_server.go:424]: yeah that's a good idea - let's add them in a separate PR once this PR is merged

File: x/staking/keeper/msg_server.go (hunk 2 above)
Comment (sampocs) [reply x/staking/keeper/msg_server.go:691]: The movement wasn't intentional. The intentional part was the bundling of a) checking if the cap is exceeded, with b) modifying progress towards the cap.

But I think moving this is also net positive cause it was a bit hidden down here at the bottom of the function

File: x/staking/keeper/liquid_stake_test.go (hunk 3 above)
Comment (sampocs) [reply x/staking/keeper/liquid_stake_test.go:58]: that module account name is arbitrary - that's not the actual format of ICA accounts.

The name here is irrelevant for the purposes of the test cause all the `AcountIsLiquidStakingProvider` function does is cast attempt to cast the account and then check the address length

File: x/staking/keeper/liquid_stake_test.go (hunk 5 above)
Comment (sampocs) [reply x/staking/keeper/liquid_stake_test.go:499]: The `DecreaseLiquidShares` function has a check that ensures the cap is enabled before it starts doing accounting changes. This function is just checking that that works.

Slight correction to your summary:
* Starts with no liquid shares and disabled cap
* Attempt to change shares, the accounting changes should not occur
* Enable cap
* Attempt to change shares again, this time the total should change

File: x/staking/keeper/liquid_stake_test.go (hunk 7 above)
Comment (sampocs) [reply x/staking/keeper/liquid_stake_test.go:435]: yeah this is just testing [this](https://github.com/iqlusioninc/liquidity-staking-module/pull/61/files/744377e109d1430f13b3572bb6cf3a686f4d6df6#diff-fb1fb56df25f422dcd246c4517169542f2a04e040d8e39d261ee8ee1477bee46R83) function. 

All that matters is that there's a validator in the store with liquid and bonded shares - no reason to complicate/bloat this test!

File: x/staking/keeper/liquid_stake_test.go (hunk 6 above)
Comment (sampocs) [reply x/staking/keeper/liquid_stake_test.go:93]: This tests [this](https://github.com/iqlusioninc/liquidity-staking-module/pull/61/files/744377e109d1430f13b3572bb6cf3a686f4d6df6#diff-fb1fb56df25f422dcd246c4517169542f2a04e040d8e39d261ee8ee1477bee46R55) function, and there's no differentiation between where the tokens came from

I do agree these are important scenarios to have covered, but I don't think this is the place to do it. It's important to keep the unit tests focused to the block of code that their testing.

Fwiw, there are more integration-like tests in `TestTokenizeSharesAndRedeemTokens` that cover these cases more closely

**Comment by sampocs** [comment]
Moved ADR to #66 so we can include a static link for the proposal

File: x/staking/keeper/liquid_stake.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (tac0turtle) [reply x/staking/keeper/liquid_stake.go:28]: in which scenarios do you see this code path getting hit? seems like if its hit there is a larger issue at hand

[the remark this answers is not in the document]
Comment (sampocs) [reply x/staking/keeper/liquid_stake.go:28]: plus, a zero would cause accounting errors which could be a bit dangerous if it was not detected (and would have to be detected with off chain monitoring)

File: adr/adr-002.md [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+    // If delegator is a liquid staking provider
+    //    Fail transaction if delegation exceeds global liquid staking cap
+    //    Fail transaction if delegation exceeds validator bond cap
+}
+
+func Undelegate() {
+    ...
+    // If the unbonded delegation is a ValidatorBond
+    //    Fail transaction if the reduction in validator bond would cause the
+    //    existing liquid delegation to exceed the cap
```
Comment (jstr1121) [inline adr/adr-002.md:307]: I think this might cause normal delegator not able to withdraw.
E.g. liquid staking global cap 30%, total delegation 100K tokens and 30K from liquid staking, 70K from normal staking.
A normal delegator who delegated 10K tokens would like to undelegate.
The delegation vs global cap changes from 30%(30K/100K) -> 33% (30K/90K) - normal users would not be able to withdraw as well if liquid staking cap riches.

File: adr/adr-002.md [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+}
+
+func BeginRedelegate() {
+    ...
+    // If the delegation is a ValidatorBond
+    //    Fail transaction if the reduction in validator bond would cause the
+    //    existing liquid delegation to exceed the cap
+
+    // If delegator is a liquid staking provider
+    //    Fail transaction if delegation exceeds global liquid staking cap
```
Comment (jstr1121) [inline adr/adr-002.md:317]: On redelegation, liquid staking cap value won't change I guess.

File: adr/adr-002.md (hunk 13 above)
Comment (sampocs) [reply adr/adr-002.md:307]: This is specifically for validator bonds (i.e. delegations that have been labeled as a validator bond), so it should have no impact on normal users unbondings

File: x/staking/keeper/liquid_stake.go [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+	// Increment the global total liquid staked
+	k.SetTotalLiquidStakedTokens(ctx, k.GetTotalLiquidStakedTokens(ctx).Add(amount))
+
+	return nil
+}
+
+// DecreaseTotalLiquidStakedTokens decrements the total liquid staked tokens
+// if the global cap is enabled
+func (k Keeper) DecreaseTotalLiquidStakedTokens(ctx sdk.Context, amount sdk.Int) {
+	if k.GlobalLiquidStakingCapEnabled(ctx) {
```
Comment (jstr1121) [inline x/staking/keeper/liquid_stake.go:100]: How does it work if global cap disabled after some time of enable and looking to enable again? I think state might be broken for this check

File: x/staking/keeper/liquid_stake.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (jstr1121) [reply x/staking/keeper/liquid_stake.go:101]: I think it would be good to add check

File: x/staking/keeper/liquid_stake.go [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+		k.SetValidator(ctx, validator)
+	}
+}
+
+// SafelyDecreaseValidatorBond decrements the total validator's self bond
+// so long as it will not cause the current delegations to exceed the threshold
+// set by validator bond factor
+func (k Keeper) SafelyDecreaseValidatorBond(ctx sdk.Context, validator types.Validator, shares sdk.Dec) error {
+	// If the cap is disabled, do nothing
+	if !k.ValidatorBondFactorEnabled(ctx) {
```
Comment (jstr1121) [inline x/staking/keeper/liquid_stake.go:139]: I think it would be helpful to count the validator bond because this could be enabled after some periods of disable time.

File: x/staking/keeper/liquid_stake.go (hunk 15 above)
Comment (sampocs) [reply x/staking/keeper/liquid_stake.go:100]: Other option is to just always count it, but my concern was that if it's disabled, we'd be wasting operations

Comment (jstr1121) [reply x/staking/keeper/liquid_stake.go:100]: I see - upgrade process would be pretty heavy if it's done on upgrade handler. (Iterating all the delegator accounts)

File: x/staking/keeper/liquid_stake.go (hunk 16 above)
Comment (sampocs) [reply x/staking/keeper/liquid_stake.go:139]: resolved in #78

File: x/staking/keeper/msg_server.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (faddat) [reply x/staking/keeper/msg_server.go:287]: I think that this could break ibc...  @vuong177 had made that as part of the splice.  Let's check with him.

**Review by zmanian — APPROVED** [review]
